src/topology/config/mod.rs

```diff
@@ -127,19 +127,37 @@ pub enum DataType {
 
 #[typetag::serde(tag = "type")]
 pub trait SourceConfig: core::fmt::Debug {
-    fn build(
-        &self,
-        name: &str,
-        globals: &GlobalOptions,
-        shutdown: ShutdownSignal,
-        out: mpsc::Sender<Event>,
-    ) -> crate::Result<sources::Source>;
+    fn build(&self, cx: SourceContext) -> crate::Result<sources::Source>;
 
     fn output_type(&self) -> DataType;
 
     fn source_type(&self) -> &'static str;
 }
 
+#[derive(Debug, Clone)]
+pub struct SourceContext<'a> {
```

Comment:
I think we should have name be `Arc<str>` and globaloptions should also be `Arc<GlobalOptions>` I don't think we ever mutate anything there beyond creation. This would allow us to remove the lifetime (which imo is not needed here).

Reply:
I actually like the explicit lifetime there... Is there a reason why you want to eliminate them?
In my view, the beauty of `&'a str` is that it's just a pointer (super cheap), while `Arc` is much more. I this case it surely doesn't matter -it's not a hot code we're talking about here - so I wouldn't be opposed to changing it to `Arc`. But I don't understand the benefits you see in switching from `&`. I see what `Arc`s allow in this case - but we don't really have a need for that flexibility, do we?

Reply:
Lifetimes make the code much more complex. One goal here is that we _don't_ need the complexity as you said because this isn't a hot path. So why go for the more complex solution? If there is someone unfamiliar with rust approaching the code base lifetimes will be scary versus encapsulating that with reference counting is much more approachable.

Reply:
I think a ref-counted global options makes a lot of sense here especially if any of the users need to keep a handle on it. That doesn't appear to be the case for either of these. Also, keeping it a reference encourages sources not to keep a copy of the whole thing, which may be useful. I'm ambivalent on this.

I'm also more sanguine about the lifetime parameter here for the big reason that it isn't leaking into the parts that most contributors will be touching—the sources themselves.

(I don't understand why Rust can do lifetime elision for references to functions, but not references in structs. If anything, that seems the more obvious rule, but that's a total digression.)

BTW, does this struct need to be `Sync`? If not, we could just use `Rc<GlobalOptions>` etc.

Reply:
It needs to be Send https://doc.rust-lang.org/stable/std/rc/struct.Rc.html#impl-Send so Rc won't work.

Reply:
An `Arc<str>` seems like it'd be more strange to a newer Rustacean than a `&'a str`. For simple data like this, I don't see a good reason for it to be ref-counted and passed all over the place.

Reply:
@lukesteensen with an `Arc<str>` internally you can expose a:

```rust
fn name(&self) -> &str
```

without the need for a lifetime on the struct.